### DISCUSSION TITLE
Render html and pdf with Julia/PlotlyLight plots

### DISCUSSION THREAD

**Discussion by dpo** [quarto-dev/quarto-cli#10569] [Q&A] [answered]
### Description

Hi. Thank you for making Quarto, which I use extensively in research and teaching.

I'm trying to render Quarto notebooks to PDF (think many reports from students, so it should be done from the command line with, e.g., `quarto render report.qmd --execute --to pdf`).
The notebooks are in Julia and produce plots with [PlotlyLight.jl](https://github.com/JuliaComputing/PlotlyLight.jl); a lightweight interface to Plotly that is important to Julia users as it produces plots much faster than any other plotting package, and so I find it rather essential in the classroom.

There's another package called [PlotlyKaleido.jl](https://github.com/JuliaPlots/PlotlyKaleido.jl) that can be used to download a plot in svg, png, pdf, etc.

The following sample notebook illustrates what I'm trying to accomplish.
The trouble is that the downloaded figure is nowhere to be found, and therefore, conversion to PDF fails.
Rending the html is broken as well in the sense that the image is not there.

    ---
    title: PlotlyLight and Quarto
    engine: julia
    ---

    ```{julia}
    #| output: false
    using Pkg
    Pkg.activate("plotlylight-quarto")
    Pkg.add("PlotlyLight")
    Pkg.add("PlotlyKaleido")

    using PlotlyLight
    using PlotlyKaleido
    PlotlyKaleido.start()
    ```

    # A plot with PlotlyLight

    ```{julia}
    p = plot(x = 1:200, y = cumsum(randn(200)), type="scatter", mode="lines" )
    p.layout.title.text = "Your stocks"
    display(p)  # this plot appears in the html rendering
    ```

    

    ```julia
    PlotlyKaleido.savefig(p, "myplot.svg")
    ```
    ![](myplot.svg)

A similar `ipynb` notebook works (the svg image is indeed downloaded and included in the html).

I opened an issue over at PlotlyKaleido.jl, but the fact that the image is downloaded with Jupyter makes me think that someone here may have an idea of what's going on.

Thank you in advance.

**Comment by mcanouil** [comment, marked as answer]
You are not executing `PlotlyKaleido.savefig(p, "myplot.svg")`.
It's a regular code block and not a code cell that is evaluated. You need the brackets.

**Reply by dpo** [reply]
DUH. Thank you. `julia` and `{julia}` are highlighted the same and I didn't spot the error.

**Comment by dpo** [comment]
Months later, I'm back to the question above. It appears that the html rendering no longer includes the plot. If I run the simplified sample notebook
````
---
title: PlotlyLight and Quarto
engine: julia
---

```{julia}
#| output: false
using Pkg
Pkg.activate("plotlylight-quarto")
Pkg.add("PlotlyLight")
Pkg.add("PlotlyKaleido")

using PlotlyLight
using PlotlyKaleido
PlotlyKaleido.start()
```

# A plot with PlotlyLight

```{julia}
p = plot(x = 1:200, y = cumsum(randn(200)), type="scatter", mode="lines" )
p.layout.title.text = "Your stocks"
display(p)  # this plot no longer appears in the html rendering
```
````
the plot pops up in a new tab of my default browser while the html file generated does not contain the plot.
I observe this behavior from VSCode and from the command line.

I am running Quarto 1.7.30 installed from Homebrew on a Mac M1.

**Reply by cscheid** [reply]
I can repro that behavior. @MichaelHatherly, can you confirm?

**Reply by cscheid** [reply]
At the end of running `quarto render` on that file, I had the following window pop up. Note the strange path:

<img width="2538" alt="image" src="https://github.com/user-attachments/assets/8dbbd914-c1b0-44e3-aed5-f27d4e2abbad" />

**Reply by MichaelHatherly** [reply]
What happens in `display(p)` is just replaced with `p`? Same thing? I'm not familiar with the internals of PlotlyLight or PlotlyKaleido. We've actively added support in QNR for https://github.com/JuliaPlots/PlotlyJS.jl. Probably is something we can add proper support for, so @dpo best open an issue over in https://github.com/PumasAI/QuartoNotebookRunner.jl.

**Reply by dpo** [reply]
NB: The 3 lines involving `PlotlyKaleido` can be removed in the script above, to further simplify it. I meant to do that but forgot, sorry.

**Reply by dpo** [reply]
> What happens in display(p) is just replaced with p? Same thing? 

Same thing.

**Reply by MichaelHatherly** [reply]
(Maybe the recent changes in https://github.com/JuliaComputing/PlotlyLight.jl/commit/a9a187016d446b16d623cd3a0032a33e3f8d18eb have something to do with this change in behaviour, if it used to work alright.)

**Reply by dpo** [reply]
You are absolutely right. The plot appears in the html output with PlotlyLight 0.9, 0.10 and 0.11, but not with 0.12. I'll open an issue there to begin with. Thank you for your assistance!

**Reply by dpo** [reply]
Actually, I'm wrong. It seems to be the interaction between PlotlyLight and PlotlyKaleido that causes the issue. If I remove the 3 lines about PlotlyKaleido in my notebook, I get the expected html output.

**Comment by dpo** [comment]
So sorry to come back again with the same problematic notebook. I'm teaching a class where the students use Quarto and, to my surprise, things that worked on macOS and Linux didn't on Windows. I personally don't use Windows, but I was able to reproduce the glitch on one of the Windows 10 workstations with a fresh install of VSCode, Julia and Quarto.
The following works fine:
`````
---
title: PlotlyLight and Quarto
engine: julia
---

```{julia}
#| output: false
using Pkg
Pkg.activate("plotlylight-quarto")
Pkg.add("PlotlyLight")
#Pkg.add("PlotlyKaleido")

using PlotlyLight
#using PlotlyKaleido
#PlotlyKaleido.start()
```

# A plot with PlotlyLight

```{julia}
p = plot(x = 1:200, y = cumsum(randn(200)), type="scatter", mode="lines" )
p.layout.title.text = "Your stocks"
display(p)  # this plot appears in the html rendering
```
`````

However, uncomment the 3 lines about `PlotlyKaleido`, and `quarto render` hangs forever, apparently stuck on the line `using Pkg`. I invoke `quarto render` from the VSCode command palette.

This time I feel that the issue is with Quarto because running the same Julia commands at a Julia prompt works fine on all platforms.

I ran `quarto check`, which came out clean. (Jupyter was missing, I installed it, and the error persists; it's not clear if installing Jupyter is required or not.)

Any ideas what may be happening here? Thanks in advance!!!

**Reply by MichaelHatherly** [reply]
Avoid doing any `Pkg` operations inside the notebook and see if that helps. Instead just create an environment using the Julia REPL and add the packages there. Then run the notebook using that environment.

**Reply by MichaelHatherly** [reply]
Maybe https://github.com/JuliaPlots/PlotlyKaleido.jl?tab=readme-ov-file#windows-note is what is being hit @dpo?

**Reply by cderv** [reply]
As I am on windows, I tested and I tried to follow the windows note but this is not working either... It hangs 

````markdown
---
title: "."
engine: julia
julia:
  exeflags: ["--project=plotlylight-quarto"]
---

# A plot with PlotlyLight

```{julia}
using PlotlyLight
```

```{julia}
using PlotlyKaleido
```

```{julia}
PlotlyKaleido.start()
```

```{julia}
p = plot(x = 1:200, y = cumsum(randn(200)), type="scatter", mode="lines" )
p.layout.title.text = "Your stocks"
display(p)  # this plot appears in the html rendering
```
````

````
❯ quartoDev preview index.qmd                     
Starting julia control server process. This might take a while...

Julia server process started.
Running [1/4] at line 10:  using PlotlyLight
Running [2/4] at line 14:  using PlotlyKaleido
Running [3/4] at line 18:  PlotlyKaleido.start()
````

Project.toml is 
````toml
[deps]
Kaleido_jll = "f7e6163d-2fa5-5f23-b69c-1db539e41963"
PlotlyKaleido = "f2990250-8cf9-495f-b13a-cce12b45703c"
PlotlyLight = "ca7969ec-10b3-423e-8d99-40f33abb42bf"
````

Even if I terminate the rendering worker is busy
````
❯ quarto call engine julia close index.qmd                                                                       
ERROR: Julia server returned error after receiving "close" command:

Failed to close notebook: C:\Users\chris\Documents\DEV_OTHER\00-TESTS\test-quarto\index.qmd

The underlying Julia error was:

Tried to close file "C:\Users\chris\Documents\DEV_OTHER\00-TESTS\test-quarto\index.qmd" but the corresponding worker is busy.
````

**Reply by dpo** [reply]
You have to use `—force` to kill quarto. I’ll try downgrading PlotlyKaleido when I have access to a Windows box later today.

**Reply by dpo** [reply]
> Avoid doing any `Pkg` operations inside the notebook and see if that helps. Instead just create an environment using the Julia REPL and add the packages there. Then run the notebook using that environment.

Thanks, but that wouldn’t be practical in a classroom.

**Reply by MichaelHatherly** [reply]
> that wouldn’t be practical in a classroom.

Just as a debugging step to work out whether it was happening during package loading or package installation.

---

Curious why it wouldn't be practical in your classroom setting? Is it the "self-contained" nature of having the `Pkg` ops in the notebook that is useful for you?

**Reply by dpo** [reply]
> Curious why it wouldn't be practical in your classroom setting? Is it the "self-contained" nature of having the Pkg ops in the notebook that is useful for you?

Yes; this is a class where (non CS) students first learn Julia and Quarto. Some are acquainted with VSCode, but not all. So in an effort to lower the barrier, I would rather not introduce any extra complications.

**Reply by dpo** [reply]
> Maybe https://github.com/JuliaPlots/PlotlyKaleido.jl?tab=readme-ov-file#windows-note is what is being hit @dpo?

@MichaelHatherly I don't think so. I just ran the same commands in a Jupyter notebook on the same Windows box, and they run just fine. Quarto continues to hang after downgrading `Kaleido_jll` as mentioned at the link that you provided.

Another hint: if I hit `Run Cell` above the Julia cell, everything goes through. It's running Quarto Preview that hangs. I hope that will put someone on the path to a solution.

Thanks!

**Reply by dpo** [reply]
In fact I don't think the issue is specific to PlotlyLight, PlotlyKaleido, etc. Quarto also hangs with certain other packages. For instance, this makes Quarto hang on Windows:
```julia
using Pkg
Pkg.activate("deblur")
Pkg.add("Images")
Pkg.add("ToeplitzMatrices")
using LinearAlgebra
using Images, ToeplitzMatrices
```

**Reply by cderv** [reply]
I don't know how QuartoNoteebookRunnner.jl does behave when activating from within a cell. 

What I tried is the following

- Create a `_setup.qmd` document to setup the Julia env for the default project (current Quarto project)
	````markdown
	---
	title: Julia env setup
	engine: julia
	---
	
	```{julia}
	using Pkg
	Pkg.add("Images")
	Pkg.add("ToeplitzMatrices")
	```
	````
- Call `quarto render _setup.qmd` so that the cell is run for installation 

- Then in main document `index.qmd` don't install package 
	````markdown
	---
	title: "test"
	engine: julia
	---
	
	## Using Julia
	
	```{julia}
	using LinearAlgebra
	using Images, ToeplitzMatrices
	```
	````

	This will use the default env and so the packages will be available. 
	
I get no hang on Windows with this process. 


I am using default project for Julia here (`--project=.`) but it can be a custom one by just adding this in the YAML for all docs
````yaml
engine: julia
julia:
  exeflags: 
    - '--project=deblur'
````

This way Julia will be started using a specific project. 


I am sharing this because it is possible that these are the installation steps using `Pkg.add` that create the handling. I'll let @MichaelHatherly, who knows way more, shares insights on this too. 

Hope it helps.

**Reply by MichaelHatherly** [reply]
> I don't know how QuartoNoteebookRunnner.jl does behave when activating from within a cell.

It's not really the way that we envisioned it being used. Our preference has been for performing all `Pkg` operations using a Julia REPL and then just pointing to already instantiated environment using the `--project` flag.

That being said, @dpo it would be best if you open an issue related to this in QNR with a reproduction of this problem and we can see what, if anything, that can be done about this.

**Reply by dpo** [reply]
> Our preference has been for performing all Pkg operations using a Julia REPL and then just pointing to already instantiated environment using the --project flag.

Really? It’s much less self contained that way. In addition, the notebooks above all work correctly on macOS and linux, so it really seems to boil down to a bug on Windows.

**Reply by MichaelHatherly** [reply]
> so it really seems to boil down to a bug on Windows.

Open an issue in https://github.com/PumasAI/QuartoNotebookRunner.jl, please.

> Really? It’s much less self contained that way.

We just have different core use cases. I'm happy to discuss ways that we can improve the "self-contained" workflow, but these cannot be done in a discussion thread in this repo. Open issues in QNR and we can discuss there instead.
